### ISSUE TITLE
Creating xmake_installer conan package - need feedback

### ISSUE THREAD

**Issue by solvingj** [xmake-io/xmake#330]
I have comments in line 29 and 31 here which show the current status.  Once we have filled those in sensibly, we will turn on CI and we will have an `xmake_installer` package for Conan.  This will then be used to test the new `conan-generator` package that we're working on. 

https://github.com/bincrafters/conan-xmake_installer/blob/2.2.3/testing/conanfile.py

Unfortunately, the instructions on the readme, and the ones used in `.travis.yml` and `appveyor.yml` are not suitable.  There's one big rule for creating these `installer` packages for Conan: 

They must not interact with the OS system files and packages: 
   - Can't use homebrew, apt/yum to install the tool
   - Can't run scripts that install things to system paths outside conan cache
   - Can't touch windows registry

The install scripts and travis/appveyor processes violate one or more of these rules, and are extremely complicated. They seem very clever, but it's VERY hard to follow what is happening. 

A build tool should be buildable with common build tools for the operating system, so Autotools/Make, or Visual Studio/nmake, or cmake is also acceptable.  Of course, it's a bit counter-intuitive to require users to use another build tool to build your new build tool.  However, this is a common scenario with various build tools, and it's the most widely accepted solution to the problem. 

Alternatively, if it's not practical for users to easily compile from source using another build tool, then we can take the route of downloading precompiled binaries.  However, the process should be consistent for all three major OS.  On Github releases page, currently, the `.exe` for windows is the only easily available binary.  There's a deb package for linux, but that installs to the OS, and doesn't work for other distros.  There does not appear to be an executable for linux, and installing via Homebrew is unacceptable for this case. 

Let me know what you think we should do!

Here are some examples of decent installer packages for other build-related tools: 
https://github.com/bincrafters/conan-ninja_installer/blob/stable/1.8.2/conanfile.py
https://github.com/bincrafters/conan-premake_installer/blob/stable/5.0.0-alpha13/conanfile.py
https://github.com/bincrafters/conan-nasm_installer/blob/stable/2.13.02/conanfile.py

**Comment by waruqi** [comment]
We can call `cd xmakesrc; ./install conandir/installdir` to install to the given directory for linux and other system. You can see https://github.com/Homebrew/homebrew-core/blob/02ca3dcde21152b0dde2c41a8cd0eae77afd2a78/Formula/xmake.rb

Or run `make build; make install conandir/installdir`

**Comment by solvingj** [comment]
Ok, is there a similar / consistent solution for windows?

**Comment by waruqi** [comment]
There is no exactly the same way on windows, compiling xmake requires an older version of xmake, and we can download the binary package and copy and install it.

https://github.com/tboox/xmake/releases/download/v2.2.3/xmake-v2.2.3.win32.zip

We can unzip it and copy to the conna install directory.

**Comment by solvingj** [comment]
As I said, it would be best if it was consistent. How did you build the first version?

**Comment by waruqi** [comment]
I built it with an older build system based on gnu makefile on msys. But I have removed the build script on windows

**Comment by waruqi** [comment]
solvingj: 
1) RE: "Why not use xmake windows binary": As I said, it's inconsistent to build Linux/macOS without having `xmake` installed, but have a different process for windows.  It is preferable to be consistent. 

2) RE: "Due to the large number of Linux distributions...".  It is possible to build a linux xmake binary that works on any linux distro.  If you are interested, I can help with this. 

3) RE: "maintaining gnu makefiles and/or nmake files".  I definitely feel bad suggesting maintaining a separate build system just to support corner cases like this one.  I understand if you choose not to do it. 

4.) I don't want to suggest anything that involves Cygwin or msys.  Perhaps we could cross-compile from Linux -> windows with MinGW though. 

In summary, again, this situation is only uncomfortable because of the combination of the following two facts: 
1. On windows, XMake cannot be built from source without downloading the binary. 
2. On Linux/macOS, XMake cannot be downloaded as a binary, and must be built from source. 

If we could change either situation, the other would cease to be relevant.

**Comment by waruqi** [comment]
> having `xmake` installed, but have a different process for windows.  It is preferable to be consistent. 
> Perhaps we could cross-compile from Linux -> windows with MinGW though. 

I agree, but I don't want to maintain a nmake files to build the windows version. Even if the build behavior is the consistent, but we need to maintain two different make files at the same time.

By using mingw to compile on linux, it is really a good choice, but still need to rely on users to install mingw toolchain.

My ultimate goal is to completely remove gnu makefiles, and use xmake to compile xmake on Linux. Of course, it is still not realistic at present. :(

Although it is very good to be able to handle it consistently, I think that for the package management tool, it should also be able to handle various platform differences, and the installation process is consistent for the user. You can see https://github.com/tboox/xmake/blob/master/xmake/repository/packages/7/7z/xmake.lua

> "Due to the large number of Linux distributions...".  It is possible to build a linux xmake binary that works on any linux distro.  If you are interested, I can help with this. 

Since I use luajit, on some distributions (e.g. centos), jit cannot be enabled, so we need compile it without jit. 

Even if we can, we still need to build at least 4 binary files at the same time, macOS x86_64, linux x86, x86_64, win x86/x64.

Maybe there may be some linux on other architecture, such as raspberry.

In summary, again, this situation is only uncomfortable because of the combination of the following two facts: 
1. On windows, XMake cannot be built from source without downloading the binary. 
2. On Linux/macOS, XMake cannot be downloaded as a binary, and must be built from source. 

If we could change either situation, the other would cease to be relevant.
